src/core/fieldformatter/qgsvaluerelationfieldformatter.h

```diff
@@ -62,23 +65,71 @@ class CORE_EXPORT QgsValueRelationFieldFormatter : public QgsFieldFormatter
 
     QVariant createCache( QgsVectorLayer *layer, int fieldIndex, const QVariantMap &config ) const override;
 
+    /**
+     * Utility to convert an array or a string representation of and array \a value to a string list
+     *
+     * \param value The value to be converted
+     * \return A string list
+     * \since QGIS 3.2
+     */
+    static QStringList valueToStringList( const QVariant &value );
+
     /**
      * Create a cache for a value relation field.
      * This can be used to keep the value map in the local memory
      * if doing multiple lookups in a loop.
+     * \param config The widget configuration
+     * \param formFeature The feature currently being edited with current attribute values
+     * \return A kvp list of values for the widget
      *
      * \since QGIS 3.0
      */
-    static QgsValueRelationFieldFormatter::ValueRelationCache createCache( const QVariantMap &config );
+    static QgsValueRelationFieldFormatter::ValueRelationCache createCache( const QVariantMap &config, const QgsFeature &formFeature = QgsFeature() );
 
     /**
-     * Utility to convert an array or a string representation of and array \a value to a string list
+     * Check if the \a expression requires a form scope (i.e. if it uses fields
+     * or geometry of the currently edited feature).
      *
-     * \param value The value to be converted
-     * \return A string list
+     * \param expression The widget's filter expression
+     * \return true if the expression requires a form scope
      * \since QGIS 3.2
      */
-    static QStringList valueToStringList( const QVariant &value );
+    static bool expressionRequiresFormScope( const QString &expression );
+
+    /**
+     * Return a list of attributes required by the form context \a expression
+     *
+     * \param expression Form filter expression
+     * \return list of attributes required by the expression
+     * \since QGIS 3.2
+     */
+    static QSet<QString> expressionFormAttributes( const QString &expression );
+
+    /**
+     * Return a list of variables required by the form context \a expression
+     *
+     * \param expression Form filter expression
+     * \return list of variables required by the expression
+     * \since QGIS 3.2
+     */
+    static QSet<QString> expressionFormVariables( const QString &expression );
+
+    /**
+     * Check whether the \a feature has all values required by the \a expression
+     *
+     * @return True if the expression can be used
+     * \since QGIS 3.2
+     */
+    static bool expressionIsUsable( const QString &expression, const QgsFeature &feature );
+
+    /**
+     * Regular expression to find dynamic filtering based on form field values
+     * \see GetCurrentFormFieldValue()
```

Comment:
GetCurrentFormFieldValue is private and hidden from the docs - so remove this line